openshift/cluster-image-registry-operator
# IR-366,IR-367,IR-411: allow users to configure private storage accounts in Azure

## Testing

This PR introduces support for private storage accounts in Azure by configuring a private endpoint when requested by the user.
Making a storage account private can happen in one of two ways:
1. user provided vnet and subnet names
2. operator discovers the vnet and subnet names

### Testing user provided vnet and subnet names
1. Provision a cluster with this PR on Azure - cluster bot example:
```
launch 4.15,openshift/cluster-image-registry-operator#930 azure
```
2. Get the vnet and subnet names used in the cluster (save it to a text file or some other accessible place):
```
oc get machinesets -nopenshift-machine-api -ojson|jq '.items[1].spec.template.spec.providerSpec.value | (.vnet, .subnet)'
```
3. Edit the registry operator config
```
oc edit configs.imageregistry/cluster
```
4. Configure the private endpoint in the registry operator config using vnet and subnet names:
```
spec:
  [...]
   storage:
      azure:
        [...]
        networkAccess:
          type: Internal
          internal:
            subnetName: <subnet-name>
            vnetName: <vnetnet-name>
```
5. Wait for the operator to configure the private endpoint (the command will return the endpoint name - this may take a couple minutes):
```
oc get configs.imageregistry/cluster -o=jsonpath="{.spec.storage.azure.privateEndpointName}" -w
```
6. Enable the default route and disable redirect on the operator config
```
oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"defaultRoute": true, "disableRedirect": true}}'
```
7. Verify that you can pull an image directly from the registry
```
REGISTRY=$(oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}')
podman login --tls-verify=false -u unused -p $(oc whoami -t) ${REGISTRY}
podman pull --tls-verify=false ${REGISTRY}/openshift/tools
```
8. Enable redirect and verify that you _cannot_ pull an image from the registry (the registry will redirect to the storage account, and the storage account is now private so this is expected to _not work_)
```
oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"disableRedirect": false}}'
# this will take a minute to propagate, check that the pods have recycled
oc get pods -nopenshift-image-registry -w
podman rmi ${REGISTRY}/openshift/tools
podman pull --tls-verify=false ${REGISTRY}/openshift/tools # this will fail with:
Error: parsing image configuration: fetching blob: StatusCode: 403, <?xml version="1.0" encoding="utf-8"?><Error><C...
```

### Testing operator discovers the vnet and subnet names
1. Provision a cluster with this PR on Azure (if you already have a cluster from the previous test you may reuse it but you need to set the registry to "Removed", wait for the transition, then set it to "Managed" again) - cluster bot example:
```
launch 4.15,openshift/cluster-image-registry-operator#930 azure
```
2. Edit the registry operator config
```
oc edit configs.imageregistry/cluster
```
3. Configure the private endpoint in the registry operator config using vnet and subnet names (note that we purposefully omit `vnetName` and `subnetName`):
```
spec:
  [...]
   storage:
      azure:
        [...]
        networkAccess:
          type: Internal
```
6. Wait for the operator to configure the private endpoint (the command will return the endpoint name - this may take a couple minutes):
```
oc get configs.imageregistry/cluster -o=jsonpath="{.spec.storage.azure.privateEndpointName}" -w
```
7. Enable the default route and disable redirect on the operator config
```
oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"defaultRoute": true, "disableRedirect": true}}'
```
8. Verify that you can pull an image directly from the registry
```
REGISTRY=$(oc get route default-route -n openshift-image-registry --template='{{ .spec.host }}')
podman login --tls-verify=false -u unused -p $(oc whoami -t) ${REGISTRY}
podman pull --tls-verify=false ${REGISTRY}/openshift/tools
```
9. Enable redirect and verify that you _cannot_ pull an image from the registry (the registry will redirect to the storage account, and the storage account is now private so this is expected to _not work_)
```
oc patch configs.imageregistry cluster --type=merge -p '{"spec":{"disableRedirect": false}}'
# this will take a minute to propagate, check that the pods have recycled
oc get pods -nopenshift-image-registry -w
podman rmi ${REGISTRY}/openshift/tools
podman pull --tls-verify=false ${REGISTRY}/openshift/tools # this will fail with:
Error: parsing image configuration: fetching blob: StatusCode: 403, <?xml version="1.0" encoding="utf-8"?><Error><C...
```
